cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/engine/SQLEngine.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api.engine;
+
+import io.cdap.cdap.api.annotation.Beta;
+import io.cdap.cdap.api.data.batch.InputFormatProvider;
+import io.cdap.cdap.api.data.batch.OutputFormatProvider;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.FromKeyValueTransform;
+import io.cdap.cdap.etl.api.PipelineConfigurable;
+import io.cdap.cdap.etl.api.SubmitterLifecycle;
+import io.cdap.cdap.etl.api.ToKeyValueTransform;
+import io.cdap.cdap.etl.api.batch.BatchContext;
+import io.cdap.cdap.etl.api.join.JoinDefinition;
+
+import java.util.concurrent.Future;
+
+/**
+ * A SQL Engine can be used to pushdown certain dataset operations.
+ *
+ * SQL Engines are implemented as plugins.
+ *
+ * Internally, the SQL Engine needs to handle retries and only surface {@link SQLEngineException} errors when there
+ * is a problem that cannot be recovered from and pipeline must be stopped.
+ *
+ * @param <KEY_OUT> The type for the Output Key when mapping a StructuredRecord
+ * @param <VALUE_OUT> The type for the Output Value when mapping a StructuredRecord
+ * @param <KEY_IN> The type for the Input Key when building a StructuredRecord
+ * @param <VALUE_IN> The type for the Input Value when building a StructuredRecord
+ */
+@Beta
+public interface SQLEngine<KEY_IN, VALUE_IN, KEY_OUT, VALUE_OUT> extends PipelineConfigurable,
+  SubmitterLifecycle<BatchContext>, FromKeyValueTransform<StructuredRecord, KEY_IN, VALUE_IN>,
+  ToKeyValueTransform<StructuredRecord, KEY_OUT, VALUE_OUT> {
+  String PLUGIN_TYPE = "sqlengine";
+
+  /**
+   * Creates an Output Format Provided that can be used to push records into a SQL Engine.
+   *
+   * After created, this dataset will be considered "locked" until the output has been committed.
+   *
+   * @param datasetName The name of the dataset to use to store these records.
+   * @param datasetSchema the schema for this dataset.
+   * @return an {@link OutputFormatProvider} instance that can be used to write records to the SQL Engine into the
+   * specified dataset.
+   */
+  OutputFormatProvider getPushProvider(String datasetName, Schema datasetSchema) throws SQLEngineException;
```

Comment:
It will be more extensible if we pass in some sort of object here (like a "PushRequest") that has a getter for the dataset name and schema, rather than passing the arguments in separately. This would allow us to add more getters in the "PushRequest" class without breaking compatibility in this API. For example, maybe we find we need to add the logical start time or something like that in the future.